### PR TITLE
Protobuf code generator (command)

### PR DESCRIPTION
**Pull request by Alma-media**
Related to https://github.com/jexia/semaphore/issues/73

```
semaphore.greeter.Say@127.0.0.1:50051> show service
+---------+-------+-----------------+------------------+
| SERVICE |  RPC  |  REQUEST TYPE   |  RESPONSE TYPE   |
+---------+-------+-----------------+------------------+
| Say     | Hello | SayHelloRequest | SayHelloResponse |
+---------+-------+-----------------+------------------+
```

```bash
$ go run main.go generate protobuf  -f ../../examples/grpc/gateway/config.hcl 
syntax = "proto2";

package semaphore.greeter;

message SayHelloRequest {
  optional string name = 1;
}

message SayHelloResponse {
  optional string msg = 1;

  optional metaNested meta = 2;

  message metaNested {
    optional int64 session = 1;
  }
}

service Say {
  rpc Hello ( SayHelloRequest ) returns ( SayHelloResponse );
}
```

### CODE REVIEW

File: cmd/semaphore/generate/protobuf/protobuf.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	if err != nil {
+		return err
+	}
+
+	collection, err := providers.Resolve(ctx, functions.Collection{}, options)
+	if err != nil {
+		return err
+	}
+
+	transporters, err := endpoints.Transporters(ctx, collection.EndpointList, collection.FlowListInterface,
```
Comment (jeroenrinzema) [inline cmd/semaphore/generate/protobuf/protobuf.go:56]: I think you should be able to pass the EndpointList as well. The `endpoints.Transporters` method is used to initialize the transporters to be passed to the listeners.

Comment (Alma-media) [reply cmd/semaphore/generate/protobuf/protobuf.go:56]: I ll check

File: pkg/transport/grpc/caller.go [hunk 2]
```diff
@@ -143,7 +148,7 @@ func (call *Call) SendMsg(ctx context.Context, rw transport.ResponseWriter, pr *
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	stream, err := grpc.NewClientStream(ctx, proxy, conn, method.fqn)
+	stream, err := grpc.NewClientStream(ctx, proxy, conn, "/"+method.String())
```
Comment (jeroenrinzema) [inline pkg/transport/grpc/caller.go:151]: maybe better to wrap the `"/"+method.String()` implementation inside a function to express the usage? ex: `method.FQN()`.

Comment (Alma-media) [reply pkg/transport/grpc/caller.go:151]: it is only a single place where we add a slash, but sure

File: cmd/semaphore/generate/protobuf/protobuf.go (hunk 1 above)
Comment (Alma-media) [reply cmd/semaphore/generate/protobuf/protobuf.go:56]: nope :disappointed:  
does not work, `grpc.ParseEndpointOptions` expects `*transport.Endpoint` to be passed

File: cmd/semaphore/generate/protobuf/protobuf.go [hunk 3]
```diff
@@ -18,22 +22,27 @@ import (
...
-	SilenceUsage: true,
-}
+	output string
+)
 
 func init() {
 	Command.PersistentFlags().StringSliceVar(&params.Protobuffers, "proto", []string{}, "If set are all proto definitions found inside the given path passed as schema definitions, all proto definitions are also passed as imports")
 	Command.PersistentFlags().StringSliceVarP(&params.Files, "file", "f", []string{"config.hcl"}, "Parses the given file as a definition file")
 	Command.PersistentFlags().StringVar(&params.LogLevel, "level", "warn", "Global logging level, this value will override the defined log level inside the file definitions")
+	Command.PersistentFlags().StringVarP(&output, "output", "o", "", "Output directory")
```
Comment (jeroenrinzema) [inline cmd/semaphore/generate/protobuf/protobuf.go:45]: Maybe it is a good idea to add a small warning that all missing directories are created?

**Pull request merged**